### ISSUE TITLE
Hub API

### ISSUE THREAD

**Issue by OR13** [decentralized-identity/secure-data-store#82]
Similar approaches:

- https://www.w3.org/TR/credential-management-1/#mediation-examples
- https://w3c-ccg.github.io/credential-handler-api

Let's assume a Chrome / Edge Extension is providing a polyfill... similar to approach taken by MetaMask.

### Feature Detection

```js
if (typeof navigator.did !== 'undefined') {
  // DID API supported.
}
```

### Grant Access to DID API

Prompt the user to allow an application to interact with the user agent via the DID API.

Similar to `ethereum.enable()` in meta mask.

```
await navigator.did.enable()
```

### Sending a message to ephemeral public keys using a service endpoint

```
const bundle = await navigator.did.hub.find({
  id: 'didcomm-ephemeral-public-key-bundle'
})
const message = await navigator.did.hub.encrypt(plaintext, bundle.publicKey);
await navigator.did.hub.send(message);
```

### Syncing changes

Needs something like: https://github.com/WoelkiM/AutoCouch/blob/master/src/__tests__/AutoCouchCRDT.test.ts

```
const todoList = await navigator.did.hub.sync({
  repo: 'todoList'
})
todoList.add('eggs');
todoList.sync();
```

**Comment by OR13** [comment]
https://arxiv.org/pdf/1912.08258.pdf

**Comment by OR13** [comment]
https://www.npmjs.com/package/bloom-filters#hyperloglog

**Comment by csuwildcat** [comment]
# Web Platform DID APIs

Assumptions:

- Some of the following APIs assume User Agent caching of DIDs and their DID Documents within the scope of a function, or from previous resolutions, with similar BF Cache and TTL characteristics.

## Core DID & DID Document Functions

### DID feature detection
```javascript=
if ('did' in navigator) {
  // Some DID APIs are supported.

  if ('datastore' in navigator.did) {
    // Personal datastore APIs are supported.
  }
}
```

### Resolve a DID

```javascript=
navigator.did.resolve('did:example:123').then(result => {

  // Your DID was resolved successfully
  console.log(result.document);

}).catch(e => {
  // DID not found, or some other error
});
```

### Request a DID from the user

```javascript=
navigator.did.requestIdentifier().then(response => {
  
  // Alice's DID: did:example:123
  console.log(response.did); 

}).catch(e => {
  // You may have been denied :/
});
```

## DID Personal Datastore APIs

### Use Case Example: *Music apps*

For the purpose of grounding the APIs described below, let's use an example of a purely clientside PWA (Progressive Web App) that plays music and curates music playlists for users. The app doesn't have any backend, it relies solely on a user's local/remote personal datastore instances.

### Opening personal datastore scopes

This function will resolve the DID to do proactive discovery of local/endpoint personal datastore instances for the DID. Think about this like preload, in Web platform terms. All subsequent API calls from the `datastore` object will be scoped to the DID passed in the `datastore.open()` invocation (similar in spirit to 'opening' an IndexedDB datastore).

```javascript=
var datastore = navigator.did.datastore.open('did:example:123');
```

### Requesting permissions to personal datastores

This API assumes whatever key, signature, or other related setup to enable the calling origin of the app to store the requested object types in the user's personal datastores are performed by the User Agent in the background.

```javascript=
var datastore = navigator.did.datastore.open('did:example:123');
datastore.requestPermissions({
  dataTypes: [
    'some-schema.info/SomeObjectType'
  ]
}).then(response => {
  // Permission granted
}).catch(e => {
  // No datastore available, or user declined request
});
```

Fully composed example, from the request of an identifier from the current user to successful permission grant:

```javascript=
const api = navigator.did;
api.requestIdentifier().then(async response => {

  try{
    var datastore = api.datastore.open(response.did);
    await datastore.requestPermissions({
      dataTypes: [
        'schema.org/MusicPlaylist',
        'schema.org/MusicRecording',
        'schema.org/AudioObject'
      ]
    });
    // Alice gave you permission!
  }
  catch(e){
    // Either Alice has no datastore available
    // or declined to let you interact with it.
  }

});
```

### Fetch all of Alice's MusicPlaylists

> Assumes we have already established a permissioned relationship with Alice's DID/datastore, as highlighted in the [request permission](#Requesting-permissions-to-personal-datastores) section above.

```javascript=
datastore.search({
  dataTypes: ['schema.org/MusicPlaylist']
}).then(response => {
  
  // Logging the returned playlists
  console.log(response.results);

})
```

### Creating a new entry for one of Alice's playlists

Objects in the system are ID'd by the hash of the whole object that is represented in the datastore. All further operations (update, delete, etc.) will be in relation to this initial create ID.

> This example assumes we have already established a permissioned relationship with Alice's DID/datastore, as highlighted in the [request permission](#Requesting-permissions-to-personal-datastores) section above. A further assumption is that whatever datastore instance(s) this create invocation is sent to automatically sync/replicate the change between each other.

```javascript=

datastore.create({
  schema: 'schema.org/MusicRecording',
  data: {
    "@type": "MusicRecording",
    "byArtist": "Lynard Skynard",
    "duration": "PT4M45S",
    "inPlaylist": "Classic Rock",
    "inAlbum": "Second Helping",
    "name": "Sweet Home Oklahoma",
    "url": "sweet-home-oklahoma"
  }
}).then(response => {
  
  // Item successfully stored.
  // Logging the hash-based ID of the object
  console.log(response.id) // 'w46ge57b5...'

})
```

### Updating a song entry from Alice's playlist

This API stores deltas in the datastore that mutate the state of objects, and changes are always made in reference to the initial create ID for a given logical object.

> This example assumes we have already established a permissioned relationship with Alice's DID/datastore, as highlighted in the [request permission](#Requesting-permissions-to-personal-datastores) section above. A further assumption is that whatever datastore instance(s) this delete invocation is sent to automatically sync/replicate the change between each other.

```javascript=

datastore.update({
  id: 'w46ge57b5...',
  data: {
    "name": "Sweet Home Alabama",
    "url": "sweet-home-alabama"
  }
}).then(response => {
  
  // Update delta successfully stored.
  // Logging the hash-based ID of the update
  console.log(response.id)

})
```

### Deleting a song entry from Alice's playlist

This example assumes we have already established a permissioned relationship with Alice's DID/datastore, as highlighted in the [request permission](#Requesting-permissions-to-personal-datastores) section above. A further assumption is that whatever datastore instance(s) this delete invocation is sent to automatically sync/replicate the change between each other.

```javascript=

datastore.deleteById(['w46ge57b5...']).then(response => {
  
  // Item successfully deleted.

})
```

**Comment by csuwildcat** [comment]
I have been hacking on this here, in case we don't want to continue posting entire API documentation entries in GH comments: https://hackmd.io/ysBkKKyoS3y7l4byDoKa6g

**Comment by OR13** [comment]
```
  await datastore.requestPermissions({
      dataTypes: [
        'schema.org/MusicPlaylist',
        'schema.org/MusicRecording',
        'schema.org/AudioObject'
      ]
    });
```

This does not seem like a well formed permissions request... If we can translate this to fit with:

https://w3c-ccg.github.io/zcap-ld/#example-2

That would make it clearer how to adopt the hub permissions with at least ZCaps...

but it worth noting that this section in particular is the most dangerous from a layer perspective... if we are not assuming a single permissions scheme, then we need a facade whereby authoriation for a logical concept can be requested in 1 format, and returned is ZCap / TXAuth / OAuth, etc.... I fear that complexity.

**Comment by jmandel** [comment]
I may be missing the point @OR13, but in the comment above you are critiquing a permission request by asking it to look more like https://w3c-ccg.github.io/zcap-ld/#example-2, which is not a request at all but a grant (a capability).

I totally agree that we need a single clear scheme for representing grants -- but I think we might have many different ways to represent requests, and different flavors of agent APIs capable of processing these squishy/variable requests and turning them into very precise grants (e.g., based on a predetermined policy, or asking the user, etc). Often a request is necessarily vague, because the requesting party wants the user to fill in some details or bind the interaction to some very specific piece of data (like for a healthcare example, a Rx refill app might ask the user to grant access to a single Rx, but the app doesn't know ahead of time which Rx to ask about, and instead relies on the user to choose one -- this keeps access narrow rather than asking the user to share their entire Rx list)... but in any case, the grant itself needs to be quite precise, and this is where I see something like ZCaps helping.

Forgive me if I'm saying something obvious or missing the point -- I'm as much trying to learn as to share a perspective :-)

**Comment by agropper** [comment]
@jmandel seems to be echoing the discussion https://github.com/decentralized-identity/secure-data-store/issues/41#issuecomment-635604197

**Comment by agropper** [comment]
See also https://github.com/w3c/did-core/issues/324

**Comment by OR13** [comment]
@jmandel yes, i guess my request is to do our best not to invent new concepts at the same time as we design a high level api. if the permissions request and permissions grants can be modeling using a system which is already in use, that will make things easier... technically, we are forbidden by the charter from inventing new authorization schemes... so my ask for this higher level layer, is to use ZCap LD or OAuth... and I have a bias against vanilla OAuth, since its not currently supported by the lower level APIs.

**Comment by agropper** [comment]
I suggest we are staying within our charter by building on https://oauth.xyz/

**Comment by OR13** [comment]
@agropper yes, because thats an authorization scheme that already exists! whereas the one proposed in this existing API does not directly use an authorization framework that already exists... although its possible that its just abstracted, and the proposed interface is using an existing authorization scheme.... its not clear, that was my point :)

**Comment by agropper** [comment]
I propose TxAuth as the **first** service endpoint to consider and then we can see if we need others. For example, we might mitigate the DID Document privacy issues by mentioning TxAuth in the core spec while leaving the door open to additional service endpoints in the future.

**Comment by OR13** [comment]
@agropper 

Lets work though TxAuth for this example then:

```
const api = navigator.did;
api.requestIdentifier().then(async response => {
  try{
    var datastore = api.datastore.open(response.did);
    await datastore.requestPermissions({
      dataTypes: [
        'schema.org/MusicPlaylist',
        'schema.org/MusicRecording',
        'schema.org/AudioObject'
      ]
    });
    // Alice gave you permission!
  }
  catch(e){
    // Either Alice has no datastore available
    // or declined to let you interact with it.
  }

});
```

Assume `navigator.did`  == `did:example:123`... with services:

```
 "service": [{
    "id":"did:example:123#tx-auth",
    "serviceEndpoint": "https://xyz.oauth.example.com"
  },{
    "id":"did:example:123#data-store",
    "serviceEndpoint": "https://data-store.example.com"
  }]
```

What does the TX Auth Request and Response look like for:

```
 await datastore.requestPermissions({
      dataTypes: [
        'schema.org/MusicPlaylist',
        'schema.org/MusicRecording',
        'schema.org/AudioObject'
      ]
    });
```

@dlongley @dmitrizagidulin can you sketch what a similar request with ZCaps would look like?

**Comment by OR13** [comment]
@tplooker I was looking at this: https://tools.ietf.org/html/draft-ietf-oauth-jwsreq-17#page-9 is this the generic version of this for OIDC?

**Comment by agropper** [comment]
@OR13 Specifically, I've raised #36 to name Access Request as the Client-> AS https://oauth.xyz/transactionrequest/ that would go the the AS service endpoint.

Also, from a data minimization perspective, I would not plan on having a 
> {
>     "id":"did:example:123#data-store",
>     "serviceEndpoint": "https://data-store.example.com"
>   }
> 
service endpoint in the DID Doc. From a privacy perspective, it would be best if the AS returned "https://data-store.example.com" to the client IFF the request is approved.

**Comment by OR13** [comment]
@agropper @csuwildcat now seems like a good time to bring this up again: https://medium.com/decentralized-identity/rhythm-and-melody-how-hubs-and-agents-rock-together-ac2dd6bf8cf4

We had a long drawn out battle over which service provider would be the only one in the services section of a did document... I suggest we not rehash that now, and instead agree that if I want to , I can update my did document however I like, and I might want to have support for 4 things, just to show that I'm interoperable...

```
 "service": [{
    "id":"did:example:123#tx-auth",
    "serviceEndpoint": "https://oauth.xyz.example.com"
  },{
    "id":"did:example:123#data-store",
    "serviceEndpoint": "https://data-store.example.com"
  },{
    "id":"did:example:123#did-comm",
    "serviceEndpoint": "https://did-comm.example.com"
  },{
    "id":"did:example:123#agent-v1",
    "serviceEndpoint": "https://agent-v1.example.com"
  }]
```

If you want to ask for about did comm, use my did comm service, if you want to ask my about tx-auth, use my tx auth service... if you want to use the hub / edv interface, use that service.

Its my did document... good luck telling me what I can and cannot put in it :)

**Comment by agropper** [comment]
Of course it's your did document @OR13 and putting photograph in it would be very convenient for both you and https://www.clearview.ai/

**Comment by OR13** [comment]
```
"service": [{
    "id":"did:example:123#my-profile-image",
    "serviceEndpoint": "https:/example.com/edvs/cfea100d-a38e-4705-b919-63b1dbf721dc/documents/z19rnJkmUqRqJTiKaR3stvnzN"
  }]
```

Assuming you have the right authorization header (OAuth Bearer / ZCap HTTP Signature / TX Auth Token)... this is what you get back:

```
{"id":"z19rnJkmUqRqJTiKaR3ftvnzN","edvId":"cbea100d-a38e-4705-b919-63b1dbf721dc","sequence":0,"indexed":[{"hmac":{"id":"urn:mockhmac:1","type":"Sha256HmacKey2019"},"sequence":0,"attributes":[{"name":"Kg8uW_vJXnALB-ng8yqxt1l2I8EBrIbB7IhKU32lCvE","value":"awrq63nVWU4yhf28PSHbxbHjiTSVY-T94_Jjocc6tgs","unique":false},{"name":"Fgi1EVWCX4Yvc-AK9X2eRDjkbh_a8R7XhvZNnwmCejA","value":"VA_8OlkMZNAL0cPDJE4zcZ1Sa7Q8-SgmjCwlEJyRduU","unique":true}]}],"jwe":{"protected":"eyJlbmMiOiJYQzIwUCJ9","recipients":[{"header":{"kid":"did:key:z6MkwJSaEMnE4u6LiqrZV1BJHSkc9x8S4mTm3ArNL1m19BZR#z6LSnHCakA8zvH61bCXv7goaT2MyVf8PjfW7XjeKTEGCgG9N","alg":"ECDH-ES+A256KW","epk":{"kty":"OKP","crv":"X25519","x":"Fdowv9z0tml4PsPRU9Xt1xeMLGnI9mylaWvp7tjZ134"},"apu":"Fdowv9z0tml4PsPRU9Xt1xeMLGnI9mylaWvp7tjZ134","apv":"ZGlkOmtleTp6Nk1rd0pTYUVNbkU0dTZMaXFyWlYxQkpIU2tjOXg4UzRtVG0zQXJOTDFtMTlCWlIjejZMU25IQ2FrQTh6dkg2MWJDWHY3Z29hVDJNeVZmOFBqZlc3WGplS1RFR0NnRzlO"},"encrypted_key":"wF42KzE4puO4MBF6Td-4rms6k49YetMZ6ryIFtX31IuDotYcSEQvPQ"}],"iv":"CG1EJ0hzysoovB6pjs4RXwlw3rq7HYf7","ciphertext":"Z_U6oBXExQPiwlxuno351WNt7anNZNfo4enyioo4Q9WDKQ9B6pNfhVYQpQUDXkPNLK6Q-upXgpqNuC4Kdfe-Igy7hzTcb7cyHTuKpDf9O0YrDl1Pq9fvZRGIsbsQESNeo46f-h6QUuyLlIco8agH1Lh-T52BpJUG1jkasHauT3CLFyASAOb5iEAWcmNvf8Z_n-0J1Iv9dvBTreO_W9QcQY-yGyi8YhMboFjcfBRf894zQSNH903axXw6J3yxti4tz0cPkdI_JNgipOlCK0I3pdqJTLEmJXGDYH11CRdJ3t0IxIIGQ1h4n3uaozk_iPVwQP4AAYLgiS8CtnexEGI4R7xvRFmrlbYDrrYJHtindbvfnRNA2x4JjlTIEUpKfx7cDa8TvUipIbbjyt7VaczQrL-qeOGIpTFKA8q58cwZSxRqT_OZXIkwL66MNMBLGk2068rWyGWpCRGvskRBYEi6Y9jKDZeNGeAyTCytX9-OpTz_aE0uBtkiC5Gh-fCocRYXuh-PutJiFOLmI2hFDoylX7cba-w1oKcfsY9wxadcEsQ1fan8P_I_VASzXFY7tB8qta_YMmyyMXa7Fw69kWw15muzH7x7SeoicJ00AJep-wsDUfYz1Grd4KFbGpNn9I1L2JW9MLUv5_XYKE-wE_Xmgln3c4keulfyRnrjSlVT79Eh9rdczTHtSt81bKzo7Aq-eRvLyrGaYLd3B3QEPAaklsGbvtrG1FR81WoHPSw_MnpkZLkl8mwcL5qvwTUNvkfP6VXy2vEgjib-txlZsDfawUPvi1rFhl99iRbiYYHeamxvtfJTOiI6Y_eUWg7iRoNhPPVqFxsKsVszZA5pPaP9YsUl0csdlzyg72m-7JnRXhh5zSAPuvWRcUIqHlF5FcWJ1T_Ct9cEhU3mX69LQVhSdVQdljWAigq6nhdJaz5SI5UbwrPu5gpvz5tCv45d44do0LAePbzJFdP4ozm0YaTKIbKCPEZYX6RI8Q3rOZb0uDmBB1pbppq0VRil9zOJ6GRSfZ51qbCw2Hwv","tag":"r5FQizQQwT6G9jqy_eCtZg"}}
```

If I choose to encrypt my profile image (or yours), clearview still needs keys to decrypt it.... and I can trust TX Auth to grant them access to my EDV, or I can hand them a capability... but getting the ciphertext and decrypting it are 2 separate steps...

As I understand TxAuth, anyone holding a bearer token that is not bound to a key can access the resource... if the resource is not encrypted, anyone holding that token can access plaintext, which might be a profile picture, or facebook friends list.

https://oauth.xyz/tokens/

TxAuth is an authorization technology, and edvs / hubs are resource servers.... edvs serve encrypted documents, but hubs might server plaintext ones... you could use TxAuth to protect the resource server, or you could use http signatures, or okta access tokens...

just like you could use ipfs, or mongo db for storage... so in summary, there is no plaintext profile images for anyone to access in an encrypted data vault, but there could be in a hub, just like my github account has a profile image... there are use cases for plaintext content.

I might use TxAuth to grant access to specific resources on my web server, or specific documents in my encrypted data vault... if those documents are not encrypted, the service provider can see them / sell them... regardless of whether i am using TxAuth or HTTP signatures... 

TxAuth is as far as I know, only about authorization... which, in the context of encrypted data vaults is a bit like a system for pulling out safe deposit boxes... regardless of how you get the box, its still locked... you need keys to open it.... in hubs, the data model is different, and maybe a hub will hand you back plaintext, but an encrypted data vault will never hand you back plaintext... because it only stores encrypted data.

To adapt the OAuth.XYZ example, I might request access to an edv docuemnt:

```
"token1": [
        {
            "actions": [
                "read",
            ],
            "locations": [
                "https://server.example.net/edvs/123/documents/456",
            ],
            "datatypes": [
                "json"
            ]
        }
    ],
```

and the AS would give me back a token:

```
 "token1": {
            "value": "OS9M2PMHKUR64TB8N6BW7OZB8CDFONP219RP1LT0",
            "type": "bearer"
        },
```

And then anyone who has possession of that token could download the ciphertext from the example above.

Or in @csuwildcat original example, some software might request access to my music list:

```
"token2": [
        {
            "actions": [
                "createPlayList",
                "updatePlayList",
                "readPlayList"
            ],
            "locations": [
                "https://hub.example.com/"
            ],
            "datatypes": [
                "data",
                "pictures",
                "music"
            ]
        }
    ]
```

and the AS would hand the software back a token:

```
"token2": {
            "value": "UFGLO2FDAFG7VGZZPJ3IZEMN21EVU71FHCARP4J1",
            "type": "bearer"
        }
```

which whoever possessed this token, would be able to "perform those hub actions"... since we don't have a clear model for how those hub actions actually work, its hard to say exactly what the client would need to submit to say: "update a music playlist" or "create a new playlist"... but if anyone get that bearer token... they will for sure be able to do those things.

TxAuth is an Authorization System, Edvs are resource servers that only serve encrypted data, and hubs are resource servers that can do all sorts of thing, if we can agree on a common way of expressing them.

**Comment by csuwildcat** [comment]
I'm seeing so many people going so many different directions on this that it's frightening to me. We should do a special topic call to lay out expectations for the singular, standard, end Hub deliverable, else I'm afraid this will spiral into a mess of 100 different mix and match configurations that fail to fulfill the fundamental goals of the effort (at least as far as I'm concerned)

**Comment by OR13** [comment]
I'd prefer to see us develop the higher level hubs api further, if it helps we can lock it down to make it simpler.

we can model the HTTP API in swagger, and assume the use of an authorization header.

We can probably just us the same tools that did:peer is using, like: https://github.com/openssi/peer-did-method-spec/issues/122#issuecomment-647001217

I think its important that we develop a shared understanding of the scope all the way from the lower levels to the highest levels, and that we have an understanding of where things like TxAuth / HTTP Signatures go... as well as how Vaults, Documents, Indexes.... Messages, Objects, etc... all look.... 

The high level api, particular around permissions is raising questions about how those are actually implemented... thats a good thing... is possible they can be implemented in more than one way... before we pick one, we need to understand those trade offs, and opportunities.... this is going to be a long process. we are going to have to be patient, and thorough, and update the spec as we go.

**Comment by tplooker** [comment]
> @tplooker I was looking at this: https://tools.ietf.org/html/draft-ietf-oauth-jwsreq-17#page-9 is this the generic version of this for OIDC?

@OR13 yes this is extracting the signed request object from oidc into its own ietf spec

**Comment by agropper** [comment]
@OR13 comment listed four service endpoint types:
```
"service": [{
    "id":"did:example:123#tx-auth",
    "serviceEndpoint": "https://oauth.xyz.example.com"
  },{
    "id":"did:example:123#data-store",
    "serviceEndpoint": "https://data-store.example.com"
  },{
    "id":"did:example:123#did-comm",
    "serviceEndpoint": "https://did-comm.example.com"
  },{
    "id":"did:example:123#agent-v1",
    "serviceEndpoint": "https://agent-v1.example.com"
  }]
```
We then added: 
```
"service": [{
    "id":"did:example:123#my-profile-image",
    "serviceEndpoint": "https:/example.com/edvs/cfea100d-a38e-4705-b919-63b1dbf721dc/documents/z19rnJkmUqRqJTiKaR3stvnzN"
  }]
```
Each of these five types is a privacy risk when crawled, indexed, and correlated without the DID controller's consent. This correlation risk is exactly the same whether or not the data itself is encrypted. What is best practice to minimize this privacy risk?

Placing an authorization service that requires an Access Request #36 ahead of each of the five types does not restrict the ability to return five different endpoints along with the authorization tokens, if any. It does not restrict encrypting some or all of those endpoints.

The problem I see with this authorization service first approach is that we don't want to choose one particular protocol such as TxAuth to the exclusion of future protocols and future transports.  This problem is not solved by encryption of the data itself.

In summary, best practice would be to have only one service endpoint per DID document and have that type specify the protocol for submitting an Access Request for private data related to that DID. We could mention TxAuth as one of the types. Additional service endpoints in the DID document would be allowed with the understanding that all service endpoints are public in the sense of correlation. These additional service endpoints could be encrypted with the goal of bypassing the Access Request endpoint and using some out-of-band means of getting the key.

**Comment by agropper** [comment]
IETF has just approved the [Grant Negotiation and Authorization Protocol (gnap)](https://datatracker.ietf.org/wg/gnap/about/) that speaks directly to the processing of an Access Request #36. The groundwork for this has been ongoing for almost two years under the TxAuth / oauth.xyz flag and the current state as of June 30, 2020 is beautifully presented in [this video](https://youtu.be/CLYB2d4akWU?t=4091). The slides are [here](https://www.dropbox.com/s/dr459qyy3t4l5yw/FHIR%20Days%20-%20XYZ.pdf?dl=0).

I urge all of us to watch the video and listen for the questions that follow. (about 35 minutes combined). Everything it describes is relevant to SDS. Although it does not mention any aspect of SSI directly, you will find plenty of ties to storage servers as related to authorization servers, DIDcomm, DIDauth and SIOP, and VC.
